plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts

```diff
@@ -0,0 +1,248 @@
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import type { HookHandlerDoneFunction } from 'fastify/types/hooks';
+import type { FastifyInstance } from 'fastify/types/instance';
+import type { FastifyReply } from 'fastify/types/reply';
+import type { FastifyRequest } from 'fastify/types/request';
+import {
+  AttributeNames,
+  FastifyNames,
+  FastifyTypes,
+} from './enums/AttributeNames';
+import type { HandlerOriginal } from './types';
+import { safeExecuteInTheMiddleMaybePromise } from './utils';
+import { VERSION } from './version';
+
+export const ANONYMOUS_NAME = 'anonymous';
+
+/** Fastify instrumentation for OpenTelemetry */
+export class FastifyInstrumentation extends InstrumentationBase {
+  constructor(config: InstrumentationConfig = {}) {
+    super(
+      '@opentelemetry/instrumentation-fastify',
+      VERSION,
+      Object.assign({}, config)
+    );
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<any>(
+        'fastify',
+        ['^3.0.0'],
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Applying patch for fastify@${moduleVersion}`);
+          return this._patchConstructor(moduleExports);
+        },
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Removing patch for fastify@${moduleVersion}`);
+        }
+      ),
+    ];
+  }
+
+  private _hookOnRequest() {
+    const instrumentation = this;
+    return function onRequest(
+      request: FastifyRequest,
+      reply: FastifyReply,
+      done: HookHandlerDoneFunction
+    ) {
+      if (!instrumentation.isEnabled()) {
+        return done();
+      }
+      const rpcMetadata = getRPCMetadata(context.active());
+      const routeName = request.routerPath;
+      if (routeName && rpcMetadata?.type === RPCType.HTTP) {
+        rpcMetadata.span.updateName(`${request.method} ${routeName || '/'}`);
+      }
+      done();
+    };
+  }
+
+  private _wrapHandler(
+    pluginName: string,
+    hookName: string,
+    original: () => Promise<any> | void
+  ): () => Promise<any> | void {
+    const instrumentation = this;
+    return function (this: any, ...args): Promise<any> | void {
+      if (!instrumentation.isEnabled()) {
+        return original.apply(this, args);
+      }
+
+      const spanName = `${FastifyNames.MIDDLEWARE} - ${
+        original.name || ANONYMOUS_NAME
+      }`;
+
+      const span = instrumentation._startSpan(spanName, {
+        [AttributeNames.FASTIFY_TYPE]: FastifyTypes.MIDDLEWARE,
+        [AttributeNames.PLUGIN_NAME]: pluginName,
+        [AttributeNames.HOOK_NAME]: hookName,
+      });
+
+      return safeExecuteInTheMiddleMaybePromise(
+        () => {
+          return original.apply(this, args);
+        },
+        err => {
+          if (err) {
+            span.setStatus({
+              code: SpanStatusCode.ERROR,
+              message: err.message,
+            });
+            span.recordException(err);
+          }
+          span.end();
```

Comment:
this function works fine for both cases, you can try yrself by running the example and modify the `client.js` instead of using `run_test` use `run_test2`. in first case it works for sync in 2nd for promises.

Reply:
Consider this example:
```js
import fastify from 'fastify';
const app = fastify({logger: true});
app.addHook('onRequest', (req, reply, done) => {
    // setTimeout is only for the example,
    // in real app this can be DB operation, api call 
    // or any other IO operation with the done callback pattern
    setTimeout(() => {
        done();
    }, 1000);
});
app.get('/', (req, reply) => {
    reply.send('hello world');
});
app.listen(3000);
```

The hook logical invocation ends when it calls the `done` function (after 1000 ms), but the generated span will end when the sync function terminates, which is after <1ms.

I suggest ending the span when the `done` function is called, and not when the function returns.

Reply:
It  works fine, the done function is called in  prehandler
```
  fastify.addHook('onRequest', (req, reply, done) => {
    const span = trace.getSpan(context.active());
    console.log('first', span);
    span.setAttribute('kuku1', 'lala');
    setTimeout(()=>{
      span.setAttribute('kuku2', 'lala');
      done();
    }, 2000);
  });

```

![image](https://user-images.githubusercontent.com/1250998/132911666-dabe2839-b7ac-4293-9c8b-9f4986c142c3.png)

Reply:
You show here the duration of the HTTP span, which is not what I am talking about.
If you look 3 rows below, you can see the span created for the middleware with the name "middleware - anonymous". This span has wrong duration of `5.818ms` instead of 2 seconds.